File: logsense/src/main/java/com/msabhi/logsense/internal/logs/LogQuery.kt
package com.msabhi.logsense.internal.logs

import com.msabhi.logsense.internal.reader.LogEntry
import com.msabhi.logsense.internal.reader.LogLevel

/**
 * Compiles an Android-Studio-style filter query into a predicate. Space-separated terms are ANDed:
 *
 * - `tag:foo`        tag contains "foo" (case-insensitive)
 * - `-tag:foo`       tag does NOT contain "foo"
 * - `message:foo`    (or `msg:foo`) message contains "foo"; `-message:` negates
 * - `level:E`        (or `level:error`) raise the minimum level
 * - `foo`            bare word: tag OR message contains "foo"
 * - `"two words"`    quotes keep spaces together
 *
 * A plain query with no `key:` is just a simple text filter, so `tag:xxx` and free text both work.
 */
internal object LogQuery {

    fun compile(filter: LogFilter): (LogEntry) -> Boolean {
        val terms = parse(filter.query)
        val min = terms.filterIsInstance<Term.MinLevel>()
            .fold(filter.minLevel) { acc, t -> maxOf(acc, t.level) }
        val predicates = terms.filterNot { it is Term.MinLevel }
        return { e -> e.level.ordinal >= min.ordinal && predicates.all { it.matches(e) } }
    }

    private fun parse(query: String): List<Term> = tokenize(query).mapNotNull { token ->
        val negate = token.length > 1 && token.startsWith('-')
        val body = if (negate) token.substring(1) else token
        val colon = body.indexOf(':')
        if (colon <= 0) return@mapNotNull if (body.isEmpty()) null else Term.Text(body, negate)
        val key = body.substring(0, colon).lowercase()
        val value = body.substring(colon + 1)
        if (value.isEmpty()) return@mapNotNull null
        when (key) {
            "tag" -> Term.Field(value, negate) { it.tag }
            "message", "msg" -> Term.Field(value, negate) { it.message }
            "level" -> LogLevel.fromName(value)?.let { Term.MinLevel(it) }
            "package", "pkg" -> null // single-app capture: package is always the host
            else -> Term.Text(body, negate) // unknown key → treat the whole token as text
        }
    }

    private fun tokenize(query: String): List<String> {
        val out = ArrayList<String>()
        val sb = StringBuilder()
        var inQuote = false
        for (c in query) {
            when {
                c == '"' -> inQuote = !inQuote
                c.isWhitespace() && !inQuote -> if (sb.isNotEmpty()) { out.add(sb.toString()); sb.setLength(0) }
                else -> sb.append(c)
            }
        }
        if (sb.isNotEmpty()) out.add(sb.toString())
        return out
    }

    private sealed interface Term {
        fun matches(e: LogEntry): Boolean

        class Field(val value: String, val negate: Boolean, val select: (LogEntry) -> String) : Term {
            override fun matches(e: LogEntry) = select(e).contains(value, ignoreCase = true) xor negate
        }

        class Text(val value: String, val negate: Boolean) : Term {
            override fun matches(e: LogEntry) =
                (e.tag.contains(value, true) || e.message.contains(value, true)) xor negate
        }

        class MinLevel(val level: LogLevel) : Term {
            override fun matches(e: LogEntry) = true
        }
    }
}

File: apple/Sources/LogSense/Internal/Logs/LogQuery.swift
import Foundation

/// A narrowing filter over the live log stream: a min level plus a query string (`LogQuery`)
/// supporting `tag:`, `-tag:`, `msg:`, `sub:`, `level:` and bare words. This is the "filter" half
/// of the filter-vs-find split; it applies to incoming lines too.
internal struct LogFilter: Equatable {
    var minLevel: LogLevel = .debug
    var query: String = ""
}

/// Compiles an Android-Studio-style filter query into a predicate. Space-separated terms are ANDed:
///
/// - `tag:foo`        tag contains "foo" (case-insensitive)
/// - `-tag:foo`       tag does NOT contain "foo"
/// - `message:foo`    (or `msg:foo`) message contains "foo"; `-message:` negates
/// - `sub:foo`        (or `subsystem:foo`) subsystem contains "foo"
/// - `level:E`        (or `level:error`) raise the minimum level
/// - `foo`            bare word: tag OR message contains "foo"
/// - `"two words"`    quotes keep spaces together
///
/// A plain query with no `key:` is just a simple text filter, so `tag:xxx` and free text both work.
internal enum LogQuery {

    static func compile(_ filter: LogFilter) -> (LogEntry) -> Bool {
        let terms = parse(filter.query)
        var min = filter.minLevel
        var predicates: [Term] = []
        for term in terms {
            if case .minLevel(let level) = term {
                min = Swift.max(min, level)
            } else {
                predicates.append(term)
            }
        }
        return { e in e.level >= min && predicates.allSatisfy { $0.matches(e) } }
    }

    private enum Term {
        case field(value: String, negate: Bool, select: (LogEntry) -> String)
        case text(value: String, negate: Bool)
        case minLevel(LogLevel)

        func matches(_ e: LogEntry) -> Bool {
            switch self {
            case .field(let value, let negate, let select):
                let hit = select(e).range(of: value, options: .caseInsensitive) != nil
                return hit != negate
            case .text(let value, let negate):
                let hit = e.tag.range(of: value, options: .caseInsensitive) != nil
                    || e.message.range(of: value, options: .caseInsensitive) != nil
                return hit != negate
            case .minLevel:
                return true
            }
        }
    }

    private static func parse(_ query: String) -> [Term] {
        tokenize(query).compactMap { token -> Term? in
            let negate = token.count > 1 && token.hasPrefix("-")
            let body = negate ? String(token.dropFirst()) : token
            guard let colon = body.firstIndex(of: ":"), colon != body.startIndex else {
                return body.isEmpty ? nil : .text(value: body, negate: negate)
            }
            let key = body[..<colon].lowercased()
            let value = String(body[body.index(after: colon)...])
            if value.isEmpty { return nil }
            switch key {
            case "tag": return .field(value: value, negate: negate) { $0.tag }
            case "message", "msg": return .field(value: value, negate: negate) { $0.message }
            case "sub", "subsystem": return .field(value: value, negate: negate) { $0.subsystem }
            case "level": return LogLevel.fromName(value).map { .minLevel($0) }
            default: return .text(value: body, negate: negate) // unknown key → treat the whole token as text
            }
        }
    }

    private static func tokenize(_ query: String) -> [String] {
        var out: [String] = []
        var current = ""
        var inQuote = false
        for c in query {
            if c == "\"" {
                inQuote.toggle()
            } else if c.isWhitespace && !inQuote {
                if !current.isEmpty { out.append(current); current = "" }
            } else {
                current.append(c)
            }
        }
        if !current.isEmpty { out.append(current) }
        return out
    }
}
